README.md
fargate-create
==============

A CLI tool for scaffolding out new [AWS ECS/Fargate](https://aws.amazon.com/fargate/) applications based on [terraform-ecs-fargate](https://github.com/turnerlabs/terraform-ecs-fargate) and [Fargate CLI](https://github.com/turnerlabs/fargate).


### Why?

The main design goal of this project is to create an easy and maintainable Fargate experience by separating `infrastructure` related concerns and `application` related concerns using tools that are optimized for each.

### Installation

You can install the CLI with a curl utility script (macos/linux) or by downloading the binary from the releases page. Once installed you'll get the `fargate-create` command.

```
curl -s get-fargate-create.turnerlabs.io | sh
````

### Usage

Assuming you have a project with a [Dockerfile]()...

Specify your template's input parameters in [terraform.tfvars](https://www.terraform.io/docs/configuration/variables.html) (or terraform.json).  The [default web application template's](https://github.com/turnerlabs/terraform-ecs-fargate) input looks something like this. Note: [any Terraform template can be used](#extensibility)

```hcl
# app/env to scaffold
app = "my-app"
environment = "dev"

internal = "true"
container_port = "8080"
replicas = "1"
health_check = "/health"
region = "us-east-1"
aws_profile = "default"
vpc = "vpc-123"
private_subnets = "subnet-123,subnet-456"
public_subnets = "subnet-789,subnet-012"
tags = {
  application = "my-app"
  environment = "dev"
}
```

```shell
$ fargate-create
scaffolding my-app dev
Looking up AWS Account ID using profile: default
downloading terraform template git@github.com:turnerlabs/terraform-ecs-fargate
installing terraform template

done
```

Now you have all the files you need to spin up something in Fargate. Note that the Terraform files can be edited or customized. You can also use your own Terraform template using the `--template` flag.

Infrastructure:  provision using Terraform
```shell
cd iac/base
terraform init && terraform apply
cd ../env/dev
terraform init && terraform apply
```

Application:  build/push using Docker and deploy using Fargate CLI
```shell
docker-compose build
login=$(aws ecr get-login --no-include-email) && eval "$login"
docker-compose push
fargate service deploy -f docker-compose.yml
```

To scaffold out additional environnments, simply change the `environment` input parameter in `terraform.tfvars` and re-run
```shell
$ fargate-create
scaffolding my-app prod
Looking up AWS Account ID using profile: default
downloading terraform template git@github.com:turnerlabs/terraform-ecs-fargate
installing terraform template
iac/base already exists, ignoring

done
```

And then bring up the new environment (no need to apply base again since it's shared):
```shell
cd ../prod
terraform init && terraform apply
```

You'll end up with a directory structure that looks something like this:
```
.
|____iac
| |____base
| |____env
| | |____dev
| | |____prod
```

As changes are made to the various upstream templates over time, the `upgrade` command can be used to keep installed versions up to date.

```bash
fargate-create upgrade
```


### Stacks

The following stacks are popular configurations that can be used with `fargate-create`

- [Web Application (ALB - HTTP or HTTPS)](https://github.com/turnerlabs/terraform-ecs-fargate#fargate-create)
- [Web Application (ALB - HTTPS/DNS)](https://github.com/turnerlabs/terraform-ecs-fargate-dns-https#fargate-create)
- [Web API Gateway](https://github.com/turnerlabs/terraform-ecs-fargate-apigateway#fargate-create)
- [Scheduled Task](https://github.com/turnerlabs/terraform-ecs-fargate-scheduled-task#fargate-create)
- [Background Worker (service)](https://github.com/turnerlabs/terraform-ecs-fargate-background-worker#fargate-create)
- [Network Application (NLB)](https://github.com/turnerlabs/terraform-ecs-fargate-nlb#fargate-create)
- [Airflow](https://github.com/turnerlabs/terraform-ecs-fargate-airflow#fargate-create)
 

### Help

```
Scaffold out new AWS ECS/Fargate applications based on Terraform templates and Fargate CLI

Usage:
  fargate-create [flags]
  fargate-create [command]

Examples:

# Scaffold an environment using the latest default template
fargate-create

# Do not prompt for options
fargate-create -y

# Use a template stored in github
fargate-create -t git@github.com:turnerlabs/terraform-ecs-fargate?ref=v0.4.3

# Scaffold out files for various build systems
fargate-create build circleciv2

# keep your template up to date
fargate-create upgrade

# Use a template stored in s3
AWS_ACCESS_KEY=xyz AWS_SECRET_KEY=xyz AWS_REGION=us-east-1 \
  fargate-create -t s3::https://s3.amazonaws.com/my-bucket/my-template

# Use a template stored in your file system
fargate-create -t ~/my-template

# Use a specific input file
fargate-create -f app.tfvars

# Use a JSON input file
fargate-create -f app.json


Available Commands:
  build       Scaffold out artifacts for various build systems
  help        Help about any command
  upgrade     Keep a terraform template up to date

Flags:
  -f, --file string         file specifying Terraform input variables, in either HCL or JSON format (default "terraform.tfvars")
  -h, --help                help for fargate-create
  -d, --target-dir string   target directory where code is outputted (default "iac")
  -t, --template string     URL of a compatible Terraform template (default "git@github.com:turnerlabs/terraform-ecs-fargate")
  -v, --verbose             Verbose output
      --version             version for fargate-create
  -y, --yes                 don't ask questions and use defaults
```


### CI/CD

Using this technique, it's easy to codegen CI/CD pipelines for many popular build tools.  The `build` command supports this.  

```shell
$ fargate-create build <provider>
```

The following providers are supported:

- [circleciv2](https://circleci.com/)
- [githubactions](https://github.com/features/actions)
- [awscodebuild](https://aws.amazon.com/codebuild/)


### Extensibility

`fargate-create` can scaffold out any Terraform template (specified by `--template`) that meets the following requirements:

- `base` and `env/dev` directory structure 
- a `env/dev/main.tf` with an s3 remote state backend
- `app` and `environment` input variables

Your template can be downloaded from a variety of locations using a variety of protocols.  The following are supported:

- Local files (`~/my-template`)
- Git (`git@github.com:my-org/my-template`)
- Amazon S3 (`s3::https://s3.amazonaws.com/my-bucket/my-template`)
- HTTP (`http://server/my-template/`)

Optionally:

- add a `fargate-create.yml` ([example here](examples/fargate-create.yml)) to your template to drive custom configuration, prompting for defaults, etc. 

An [example](https://github.com/turnerlabs/terraform-ecs-fargate-scheduled-task/) of an extended template:
```shell
$ fargate-create -f my-scheduledtask.tfvars -t git@github.com:turnerlabs/terraform-ecs-fargate-scheduled-task
```


CHANGELOG.md
## 0.9.0 (2020-07-06)

### Features

- adds support for aws cli v2 when generating the deploy.sh script (#34)


## 0.8.0 (2019-12-09)

### Features

- adds `githubactions` build provider [#32](https://github.com/turnerlabs/fargate-create/pull/32)
- adds `awscodebuild` build provider [#33](https://github.com/turnerlabs/fargate-create/pull/33)


## 0.7.1 (2019-08-12)

### Bug Fixes

- AWS Region not respected for tf state bucket in environments [#27](https://github.com/turnerlabs/fargate-create/issues/27)


## 0.7.0 (2019-06-10)

### Features

- updates outputted docker-compose.yml to v3.7


## 0.6.1 (2019-02-22)

### Features

- Documentation


## 0.6.0 (2019-02-22)

### Features

- New `upgrade` command ([#20](https://github.com/turnerlabs/fargate-create/issues/20))


## 0.5.0 (2019-01-09)

### Features

- New `build` command framework with support for Circle CI ([#16](https://github.com/turnerlabs/fargate-create/pull/16))
- Release binaries now hosted in s3 ([#18](https://github.com/turnerlabs/fargate-create/pull/18))


## 0.4.0 (2018-11-14)

### Features

- adds support for scheduled tasks ([#15](https://github.com/turnerlabs/fargate-create/pull/15))


## 0.3.0 (2018-09-20)

### Features

- Sets optional port in docker-compose.yml ([#13](https://github.com/turnerlabs/fargate-create/issues/13))

### Bug Fixes

  - Fixes issue parsing Terraform comments ([#12](https://github.com/turnerlabs/fargate-create/issues/12))


## 0.2.1 (2018-09-12)

### Features

  - Readme doc update


## 0.2.0 (2018-08-21)

### Features

  - Changed the default template to always pull the latest (using git)


## 0.1.0 (2018-08-08)

### Features

  - Initial Release


cmd/terraform_test.go
package cmd

import (
	"fmt"
	"strings"
	"testing"
)

func TestUpdateTerraformBackend(t *testing.T) {

	tf := `
terraform {
	backend "s3" {
		region  = "us-east-1"
		profile = ""
		bucket  = ""
		key     = "dev.terraform.tfstate"
	}
}

provider "aws" {
	region  = "${var.region}"
	profile = "${var.aws_profile}"
}
`

	profile := "my-profile"
	app := "my-app"
	env := "qa"
	region := "us-east-1"
	result := updateTerraformBackend(tf, profile, app, env, region)
	t.Log(result)

	expected := fmt.Sprintf(`profile = "%s"`, profile)
	if !strings.Contains(result, expected) {
		t.Errorf("expected: %s; actual: %s", expected, result)
	}
	expected = fmt.Sprintf(`bucket  = "tf-state-%s"`, app)
	if !strings.Contains(result, expected) {
		t.Errorf("expected: %s; actual: %s", expected, result)
	}
}

func TestUpdateTerraformBackend_Region(t *testing.T) {

	tf := `
terraform {
	backend "s3" {
		region  = "us-east-1"
		profile = ""
		bucket  = ""
		key     = "dev.terraform.tfstate"
	}
}

provider "aws" {
	region  = "${var.region}"
	profile = "${var.aws_profile}"
}
`

	profile := "my-profile"
	app := "my-app"
	env := "qa"
	region := "us-west-1"
	result := updateTerraformBackend(tf, profile, app, env, region)
	t.Log(result)

	expected := fmt.Sprintf(`profile = "%s"`, profile)
	if !strings.Contains(result, expected) {
		t.Errorf("expected: %s; actual: %s", expected, result)
	}
	expected = fmt.Sprintf(`bucket  = "tf-state-%s"`, app)
	if !strings.Contains(result, expected) {
		t.Errorf("expected: %s; actual: %s", expected, result)
	}
	expected = fmt.Sprintf(`region  = "%s"`, region)
	if !strings.Contains(result, expected) {
		t.Errorf("expected: %s; actual: %s", expected, result)
	}
}


func TestParseInputVars(t *testing.T) {

	tf := `
	region = "us-east-2"

	aws_profile = "default"
	
	saml_role = "devops"
	
	app = "my-app"
	
	environment = "qa"

	container_port = "8080"
	
	tags = {
		application   = ""
		environment   = "dev"
		team          = ""
		customer      = ""
		contact-email = ""
	}
	
	internal = "true"
`
	app, env, profile, region, containerPort, err := parseInputVars(varFormatHCL, tf)

	t.Log(app)
	t.Log(env)
	t.Log(profile)
	t.Log(region)
	t.Log(containerPort)
	if err != nil {
		t.Error(err)
	}

	expected := "my-app"
	if app != expected {
		t.Errorf("expected: %s; actual: %s", expected, app)
	}
	expected = "qa"
	if env != expected {
		t.Errorf("expected: %s; actual: %s", expected, env)
	}
	expected = "default"
	if profile != expected {
		t.Errorf("expected: %s; actual: %s", expected, profile)
	}
	expected = "us-east-2"
	if region != expected {
		t.Errorf("expected: %s; actual: %s", expected, profile)
	}
	expected = "8080"
	if containerPort != expected {
		t.Errorf("expected: %s; actual: %s", containerPort, profile)
	}
}

func TestParseInputVars_OptionalContainerPort(t *testing.T) {

	tf := `
	region = "us-east-2"

	aws_profile = "default"
	
	saml_role = "devops"
	
	app = "my-app"
	
	environment = "qa"

	tags = {
		application   = ""
		environment   = "dev"
		team          = ""
		customer      = ""
		contact-email = ""
	}
	
	internal = "true"
`
	app, env, profile, region, containerPort, err := parseInputVars(varFormatHCL, tf)

	t.Log(app)
	t.Log(env)
	t.Log(profile)
	t.Log(region)
	t.Log(containerPort)
	if err != nil {
		t.Error(err)
	}

	expected := "my-app"
	if app != expected {
		t.Errorf("expected: %s; actual: %s", expected, app)
	}
	expected = "qa"
	if env != expected {
		t.Errorf("expected: %s; actual: %s", expected, env)
	}
	expected = "default"
	if profile != expected {
		t.Errorf("expected: %s; actual: %s", expected, profile)
	}
	expected = "us-east-2"
	if region != expected {
		t.Errorf("expected: %s; actual: %s", expected, profile)
	}
	expected = ""
	if containerPort != expected {
		t.Errorf("expected: %s; actual: %s", containerPort, profile)
	}
}

func TestParseInputVars_comments(t *testing.T) {

	tf := `
	region = "us-east-2"

	aws_profile = "default"
	
	saml_role = "devops"
	
	app = "my-app" #comment
	
	environment = "qa"
	
	tags = {
		application   = ""
		environment   = "dev"
		team          = ""
		customer      = ""
		contact-email = ""
	}
	
	internal = "true"
`
	app, env, profile, region, _, err := parseInputVars(varFormatHCL, tf)

	t.Log(app)
	t.Log(env)
	t.Log(profile)
	t.Log(region)
	if err != nil {
		t.Error(err)
	}

	expected := "my-app"
	if app != expected {
		t.Errorf("expected: %s; actual: %s", expected, app)
	}
	expected = "qa"
	if env != expected {
		t.Errorf("expected: %s; actual: %s", expected, env)
	}
	expected = "default"
	if profile != expected {
		t.Errorf("expected: %s; actual: %s", expected, profile)
	}
	expected = "us-east-2"
	if region != expected {
		t.Errorf("expected: %s; actual: %s", expected, profile)
	}
}

func TestParseInputVars_JSON(t *testing.T) {

	tf := `{
  "region": "us-east-2",
	"aws_profile": "default",	
	"saml_role": "devops",	
	"app": "my-app",	
	"environment": "qa",	
	"container_port": "8080",
	"tags": {
		"application": "",
		"environment": "dev",
		"team": "",
		"customer": "",
		"contact-email": ""
	},	
	"internal": true
}
`
	app, env, profile, region, containerPort, err := parseInputVars(varFormatJSON, tf)

	t.Log(app)
	t.Log(env)
	t.Log(profile)
	t.Log(region)
	t.Log(containerPort)
	if err != nil {
		t.Error(err)
	}

	expected := "my-app"
	if app != expected {
		t.Errorf("expected: %s; actual: %s", expected, app)
	}
	expected = "qa"
	if env != expected {
		t.Errorf("expected: %s; actual: %s", expected, env)
	}
	expected = "default"
	if profile != expected {
		t.Errorf("expected: %s; actual: %s", expected, profile)
	}
	expected = "us-east-2"
	if region != expected {
		t.Errorf("expected: %s; actual: %s", expected, profile)
	}
	expected = "8080"
	if containerPort != expected {
		t.Errorf("expected: %s; actual: %s", expected, profile)
	}

}


cmd/scaffolder_test.go
package cmd

import (
	"io/ioutil"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

const tmpDir = "./tmptest"

func TestIgnoreSensitiveFiles_New(t *testing.T) {

	//arrange
	teardownTestCase := setupTestCase(t)
	defer teardownTestCase(t)
	file := filepath.Join(tmpDir, ".gitignore")

	//act
	ignored := []string{"foo"}
	ensureFileContains(file, ignored)

	//assert
	//fail if file doesn't contain 1 instance of foo
	dat, err := ioutil.ReadFile(file)
	if err != nil {
		t.Error(err)
	}
	contents := string(dat)
	t.Log(contents)
	if strings.Count(contents, "foo") != 1 {
		t.Error("expecting 1 occurance of foo")
	}
}

func TestIgnoreSensitiveFiles_Existing(t *testing.T) {

	//arrange
	teardownTestCase := setupTestCase(t)
	defer teardownTestCase(t)
	file := filepath.Join(tmpDir, ".gitignore")

	//create .gitignore with existing entry
	d1 := []byte("foo\nbar")
	err := ioutil.WriteFile(file, d1, 0644)
	if err != nil {
		t.Error(err)
	}

	//test
	ignored := []string{"foo"}
	ensureFileContains(file, ignored)

	//assert
	//fail if file doesn't contain 1 instance of foo
	dat, err := ioutil.ReadFile(file)
	if err != nil {
		t.Error(err)
	}
	contents := string(dat)
	t.Log(contents)
	if strings.Count(contents, "foo") != 1 {
		t.Error("expecting 1 occurance of foo")
	}
}

func TestIgnoreSensitiveFiles_Existing_Multiple(t *testing.T) {

	//arrange
	teardownTestCase := setupTestCase(t)
	defer teardownTestCase(t)
	file := filepath.Join(tmpDir, ".gitignore")

	//create .gitignore with existing entry
	d1 := []byte("foo\nbar")
	err := ioutil.WriteFile(file, d1, 0644)
	if err != nil {
		t.Error(err)
	}

	//test
	ignored := []string{"foo", "baz"}
	ensureFileContains(file, ignored)

	//assert
	//fail if file doesn't contain 1 instance of foo
	dat, err := ioutil.ReadFile(file)
	if err != nil {
		t.Error(err)
	}
	contents := string(dat)
	t.Log(contents)
	if strings.Count(contents, "foo") != 1 {
		t.Error("expecting 1 occurance of foo")
	}
	if strings.Count(contents, "bar") != 1 {
		t.Error("expecting 1 occurance of bar")
	}
	if strings.Count(contents, "baz") != 1 {
		t.Error("expecting 1 occurance of baz")
	}
}

func setupTestCase(t *testing.T) func(t *testing.T) {
	t.Log("setup test case")

	//create tmpDir
	err := os.MkdirAll(tmpDir, 0755)
	if err != nil {
		t.Error("error creating", tmpDir)
	}

	return func(t *testing.T) {
		t.Log("teardown test case")

		//delete
		os.RemoveAll(tmpDir)
	}
}

func TestDockerComposeYml_Port(t *testing.T) {

	context := scaffoldContext{
		AccountID:     "abc",
		App:           "my-app",
		ContainerPort: "3000",
		Env:           "dev",
		Profile:       "default",
		Region:        "us-east-1",
		Format:        ".tfvars",
	}

	yml := getDockerComposeYml(&context)
	t.Log(yml)

	expected := "- 3000:3000"
	if !strings.Contains(yml, expected) {
		t.Errorf("expected: %s; actual: %s", expected, yml)
	}
}

func TestDockerComposeYml_NoPort(t *testing.T) {

	context := scaffoldContext{
		AccountID:     "abc",
		App:           "my-app",
		ContainerPort: "",
		Env:           "dev",
		Profile:       "default",
		Region:        "us-east-1",
		Format:        ".tfvars",
	}

	yml := getDockerComposeYml(&context)
	t.Log(yml)

	notexpected := "ports:"
	if strings.Contains(yml, notexpected) {
		t.Errorf("not expected: %s; actual: %s", notexpected, yml)
	}	
}


cmd/build.go
package cmd

import (
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"

	"github.com/spf13/cobra"
	"github.com/turnerlabs/fargate-create/cmd/build"
)

var buildCmd = &cobra.Command{
	Use:   "build",
	Short: "Scaffold out artifacts for various build systems",
	Long:  "Scaffold out artifacts for various build systems",
	Args:  cobra.ExactArgs(1),
	Run:   doBuild,
	Example: `
fargate-create build local
fargate-create build circleciv2
fargate-create build githubactions
`,
}

func init() {
	rootCmd.AddCommand(buildCmd)
}

func doBuild(cmd *cobra.Command, args []string) {

	//load build provider
	providerString := args[0]
	provider, err := build.GetProvider(providerString)
	check(err)

	//get artifacts
	artifacts, err := provider.ProvideArtifacts(context)
	check(err)

	//write artifacts to file system
	if artifacts != nil {
		for _, artifact := range artifacts {
			//create directories if needed
			dirs := filepath.Dir(artifact.FilePath)
			err = os.MkdirAll(dirs, os.ModePerm)
			check(err)

			if _, err := os.Stat(artifact.FilePath); err == nil {
				//exists
				fmt.Print(artifact.FilePath + " already exists. Overwrite? ")
				if askForConfirmation() {
					err = ioutil.WriteFile(artifact.FilePath, []byte(artifact.FileContents), artifact.FileMode)
					fmt.Println("wrote " + artifact.FilePath)
					check(err)
				}
			} else {
				//doesn't exist
				err = ioutil.WriteFile(artifact.FilePath, []byte(artifact.FileContents), artifact.FileMode)
				fmt.Println("wrote " + artifact.FilePath)
				check(err)
			}
		}
	}

}


cmd/scaffolder.go
package cmd

import (
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"strings"

	getter "github.com/hashicorp/go-getter"
	yaml "gopkg.in/yaml.v2"
)

type scaffoldTemplate struct {
	Base templateDirectory
	Env  templateDirectory
}

type templateDirectory struct {
	Directory     string
	Configuration *templateConfig
	Installed     bool
}

const templateTypeService = "Service"
const templateTypeScheduledTask = "ScheduledTask"
const defaultTemplateType = templateTypeService
const baseDir = "base"
const envDir = "env"
const devDir = "dev"

type templateConfig struct {
	TemplateType string    `yaml:"templateType"`
	Prompts      []*prompt `yaml:"prompts"`
}

type prompt struct {
	Question          string   `yaml:"question"`
	Default           string   `yaml:"default"`
	FilesToDeleteIfNo []string `yaml:"filesToDeleteIfNo"`
}

func scaffold(context *scaffoldContext) {

	//scaffold out infrastructure files
	template := scaffoldInfrastructure(context)

	//scaffold application files
	scaffoldApplication(context, template)

	//apply any template configurations
	applyTemplateConfiguration(template.Base)
	applyTemplateConfiguration(template.Env)
}

func applyTemplateConfiguration(t templateDirectory) {
	if t.Configuration != nil {
		for _, prompt := range t.Configuration.Prompts {
			//if -y, use defaults, otherwise prompt
			response := prompt.Default
			if !yesUseDefaults {
				fmt.Println()
				q := fmt.Sprintf("%s (%s) ", prompt.Question, prompt.Default)
				response = promptAndGetResponse(q, prompt.Default)
			}
			yes := containsString(okayResponses, response)
			if !yes && prompt.FilesToDeleteIfNo != nil {
				for _, file := range prompt.FilesToDeleteIfNo {
					p := filepath.Join(t.Directory, file)
					fmt.Println("deleting ", p)
					err := os.Remove(p)
					check(err)
				}
			}
		}
	}
}

func scaffoldInfrastructure(context *scaffoldContext) *scaffoldTemplate {

	//fetch terraform template
	templateDir := downloadTerraformTemplate()
	debug("downloaded to:", templateDir)

	result := installTerraformTemplate(templateDir, context.Env)
	debug("environment installed to:", result.Env.Directory)

	//copy var file into base module
	if result.Base.Installed {
		debug(fmt.Sprintf("copying %s to %s", varFile, result.Base.Directory))
		targetFile := getTargetVarFile(context.Format)
		err := copyFile(varFile, filepath.Join(result.Base.Directory, targetFile))
		check(err)
	}

	//copy var file into environment module
	debug(fmt.Sprintf("copying %s to %s", varFile, result.Env.Directory))
	targetFile := getTargetVarFile(context.Format)
	err := copyFile(varFile, filepath.Join(result.Env.Directory, targetFile))
	check(err)

	//update tf backend in main.tf to match app/env
	transformMainTFToContext(result.Env.Directory, context.Profile, context.App, context.Env, context.Region)

	return result
}

func transformMainTFToContext(dir string, profile string, app string, env string, region string) {
	mainTfFile := filepath.Join(dir, "main.tf")
	fileBits, err := ioutil.ReadFile(mainTfFile)
	check(err)
	maintf := updateTerraformBackend(string(fileBits), profile, app, env, region)
	err = ioutil.WriteFile(mainTfFile, []byte(maintf), 0644)
	check(err)
}

//fetches and installs the tf template and returns the output directory
func downloadTerraformTemplate() string {

	client := getter.Client{
		Src:  templateURL,
		Dst:  "./" + tempDir,
		Mode: getter.ClientModeAny,
	}

	fmt.Println("downloading terraform template", templateURL)
	err := client.Get()
	check(err)
	debug("done")

	return tempDir
}

//installs a template for the specified environment and returns a scaffoldTemplate
func installTerraformTemplate(templateDir string, environment string) *scaffoldTemplate {

	result := scaffoldTemplate{
		Base: templateDirectory{},
		Env:  templateDirectory{},
	}

	//create infrastructure directory (if not already there)
	targetInfraDir := targetDir
	fmt.Println("installing terraform template")
	if _, err := os.Stat(targetInfraDir); os.IsNotExist(err) {
		debug("creating directory:", targetInfraDir)
		err = os.MkdirAll(targetInfraDir, 0755)
		check(err)
	} else {
		debug(targetInfraDir + " already exists")
	}

	//copy over infrastructure/base (if not already there)
	sourceBaseDir := filepath.Join(templateDir, baseDir)
	destBaseDir := filepath.Join(targetInfraDir, baseDir)
	if _, err := os.Stat(destBaseDir); os.IsNotExist(err) {
		debug(fmt.Sprintf("copying %s to %s", sourceBaseDir, destBaseDir))
		err = copyDir(sourceBaseDir, destBaseDir)
		check(err)

		result.Base.Installed = true
		result.Base.Directory = destBaseDir

		//does template contain a fargate-create.yml config?  is so, load it
		result.Base.Configuration = loadTemplateConfig(result.Base.Directory)

	} else {
		fmt.Println(destBaseDir + " already exists, ignoring")
	}

	//if environment directory exists, prompt to override, if no, then exit
	sourceEnvDir := filepath.Join(templateDir, envDir, devDir)
	destEnvDir := filepath.Join(targetInfraDir, envDir, environment)

	yes := true
	if _, err := os.Stat(destEnvDir); err == nil {
		//exists
		fmt.Print(destEnvDir + " already exists. Overwrite? ")
		if yes = askForConfirmation(); yes {
			debug("deleting", destEnvDir)
			//delete environment directory (all files)
			err = os.RemoveAll(destEnvDir)
			check(err)
		}
	} else {
		//doesn't exist
		debug(destEnvDir + " doesn't exist")
	}

	if yes {
		//env directory either doesn't exist or user wants to overwrite
		//copy repo/env/${env} -> ./infrastructure/env/${env}
		debug(fmt.Sprintf("copying %s to %s", sourceEnvDir, destEnvDir))
		err := copyDir(sourceEnvDir, destEnvDir)
		check(err)

		result.Env.Installed = true
		result.Env.Directory = destEnvDir

		//does template contain a fargate-create.yml config?  is so, load it
		result.Env.Configuration = loadTemplateConfig(result.Env.Directory)
	}

	// finally, delete temp dir
	debug("deleting:", tempDir)
	err := os.RemoveAll(tempDir)
	check(err)

	return &result
}

func loadTemplateConfig(dir string) *templateConfig {
	configFile := filepath.Join(dir, templateConfigFile)
	var config templateConfig
	if _, err := os.Stat(configFile); !os.IsNotExist(err) {
		debug("found template config: ", dir)
		//load yaml
		dat, err := ioutil.ReadFile(configFile)
		check(err)
		err = yaml.Unmarshal(dat, &config)
		check(err)

		//default template type
		if config.TemplateType == "" {
			config.TemplateType = defaultTemplateType
		}
	} else {
		debug("didn't find template config: ", dir)
		return nil
	}
	return &config
}

func getTargetVarFile(format string) string {
	targetFile := ""
	if format == varFormatHCL {
		targetFile = "terraform.tfvars"
	}
	if format == varFormatJSON {
		targetFile = "terraform.tfvars.json"
	}
	return targetFile
}

func scaffoldApplication(context *scaffoldContext, t *scaffoldTemplate) {

	//write the application files to the env directory
	targetAppDir := t.Env.Directory

	//write a docker-compose.yml file
	dockerComposeYml := getDockerComposeYml(context)
	dockerComposeYmlFile := filepath.Join(targetAppDir, "docker-compose.yml")
	debug("writing", dockerComposeYmlFile)
	err := ioutil.WriteFile(dockerComposeYmlFile, []byte(dockerComposeYml), 0644)
	check(err)

	//write hidden.env
	hiddenEnvFileName := filepath.Join(targetAppDir, "hidden.env")
	sampleContents := "#FOO=bar\n"
	err = ioutil.WriteFile(hiddenEnvFileName, []byte(sampleContents), 0644)
	check(err)

	//write a fargate.yml for the cli
	fargateYml := getFargateYaml(context, t.Env.Configuration)
	fargateYmlFile := filepath.Join(targetAppDir, "fargate.yml")
	debug("writing", fargateYmlFile)
	err = ioutil.WriteFile(fargateYmlFile, []byte(fargateYml), 0644)
	check(err)

	//write deploy.sh
	deployScript := getDeployScript(context, t.Env.Configuration)
	deployScriptFile := filepath.Join(targetAppDir, "deploy.sh")
	debug("writing", deployScriptFile)
	err = ioutil.WriteFile(deployScriptFile, []byte(deployScript), 0755)
	check(err)

	//ignored files
	hiddenenv := strings.Split(hiddenEnvFileName, "/")
	ignoredFiles := []string{hiddenenv[len(hiddenenv)-1], ".terraform"}
	ensureFileContains(".gitignore", ignoredFiles)
	ensureFileContains(".dockerignore", ignoredFiles)
}

func getFargateYaml(context *scaffoldContext, config *templateConfig) string {
	textTemplate := `cluster: {{.App}}-{{.Env}}
service: {{.App}}-{{.Env}}`
	if config.TemplateType == templateTypeScheduledTask {
		textTemplate += `
task: {{.App}}-{{.Env}}
rule: {{.App}}-{{.Env}}
`
	}
	return applyTemplate(textTemplate, context)
}

func getDockerComposeYml(context *scaffoldContext) string {

	//only add docker ports if container_port context variable is specified
	var t string
	if context.ContainerPort != "" {
		t = `version: "3.7"
services:
  {{.App}}:
    build: ../../../
    image: {{.AccountID}}.dkr.ecr.{{.Region}}.amazonaws.com/{{.App}}:0.1.0
    ports:
    - {{.ContainerPort}}:{{.ContainerPort}}
    env_file:
    - hidden.env
    labels:
      aws.ecs.fargate.deploy: 1
    #x-fargate-secrets:
      #KEY: value
`
	} else {
		t = `version: "3.7"
services:
  {{.App}}:
    build: ../../../
    image: {{.AccountID}}.dkr.ecr.{{.Region}}.amazonaws.com/{{.App}}:0.1.0
    env_file:
    - hidden.env
    labels:
      aws.ecs.fargate.deploy: 1
    #x-fargate-secrets:
      #KEY: value
`
	}
	return applyTemplate(t, context)
}

func getDeployScript(context *scaffoldContext, config *templateConfig) string {
	t := `#!/bin/bash
set -e

# build image
docker-compose build

export AWS_PROFILE={{.Profile}}
export AWS_DEFAULT_REGION={{.Region}}

# login to ECR
version=$(aws --version | awk -F'[/.]' '{print $2}')
if [ $version -eq "1" ]; then
  login=$(aws ecr get-login --no-include-email) && eval "$login"
else
  aws ecr get-login-password | docker login --username AWS --password-stdin {{.AccountID}}.dkr.ecr.{{.Region}}.amazonaws.com
fi

# push image to ECR repo
docker-compose push`

	if config.TemplateType == templateTypeService {
		t += `
# deploy image and env vars
fargate service deploy -f docker-compose.yml
`
	} else if config.TemplateType == templateTypeScheduledTask {
		t += `
# deploy image and env vars
REVISION=$(fargate task register -f docker-compose.yml)
fargate events target -r ${REVISION}
`
	}

	return applyTemplate(t, context)
}
